File: Maze-Generator/Maze.cpp
#include <iostream>
#include <string>

#include <cstdlib>
#include <ctime>

#include "Maze.h"

namespace mazeGenerator {

  using std::string;

  //===========================================================================
  // Constructor
  //
  // Receives an integer with the width/height of the board and initializes
  // the object
  //===========================================================================
  Maze::Maze (int size) {
    std::srand ((unsigned int) std::time (NULL));

    init (size);
  }


  //===========================================================================
  // Destructor
  //===========================================================================
  Maze::~Maze () {
    delete [] _link;
    delete [] _board;
  }


  //===========================================================================
  // init
  // 
  // This method initializes the attributes for the class, receives the width
  // of the board and sets its total size
  //===========================================================================
  void Maze::init (int size) {
    if (size > 101 || size < 7 || size % 2 == 0) {
      std::cerr << "Board size not correctly set. It must be an odd number between 7 and 101." << std::endl;
    } else {
      _numberLinks = 0;
      _size = (size - 1) / 2;
      _totalSize = _size * _size;

      // Initializing the board
      _board = new char [_totalSize];

      // Filling the board with walls
      for (int i = 0; i < _totalSize; _board [i] = BOARD_WALL, i++);

      // Placing the starting point
      _board [0] = BOARD_START;

      // Placing the finish point
      _board [_totalSize - 1] = BOARD_FINISH;

      create ();
    }
  }


  //===========================================================================
  // create
  // 
  // Creates the maze
  //===========================================================================
  void Maze::create () {
    bool foundFinish = false;

    // Looping while there still are breakable walls in the maze
    do {

      bool hasAvailableCell;
      int currentCell = getCellWithWallParent (!foundFinish);

      // Looping to expand the path until it cannot be expanded anymore
      do {
        int selectedCell = getRandomWallParent (currentCell, !foundFinish);

        // If there is an available cell next to the current one, jumps to it
        if (hasAvailableCell = selectedCell > -1) {

          // If it reached the finish, stops this path and starts a new one
          if (selectedCell == _totalSize - 1) {
            foundFinish = true;
            hasAvailableCell = false;
          } else {
            _board [selectedCell] = BOARD_EMPTY;
          }

          pushLink (currentCell, selectedCell);

          currentCell = selectedCell;
        }

      } while (hasAvailableCell);

    } while (getCellWithWallParent (!foundFinish) > -1);
  }


  //===========================================================================
  // pushLink
  // 
  // Creates a link between 2 cells and pushes it in the array of links
  //===========================================================================
  void Maze::pushLink (int cell1, int cell2) {
    BoardLink newLink (cell1, cell2);

    // Craeting a new link bundle
    BoardLink* newBundle = new BoardLink [_numberLinks + 1];

    // Copying the previous links to the new bundle
    for (int i = 0; i < _numberLinks; newBundle [i] = _link [i], i++);

    // Adding the new link
    newBundle [_numberLinks] = newLink;

    // Replacing the previous bundle
    delete [] _link;
    _link = newBundle;

    _numberLinks++;
  }


  //===========================================================================
  // hasLink
  // 
  // Creates a link between 2 cells and pushes it in the array of links
  //===========================================================================
  bool Maze::hasLink (int cell1, int cell2) {
    bool hasLink = false;

    // Looping through the bundle looking for the link
    for (int i = 0; i < _numberLinks && !hasLink; i++) {
      bool condition1 = _link [i].c1 == cell1 && _link [i].c2 == cell2;
      bool condition2 = _link [i].c1 == cell2 && _link [i].c2 == cell1;

      if (condition1 || condition2) hasLink = true;
    }

    return hasLink;
  }


  //===========================================================================
  // getCellWithWallParent
  // 
  // Loops through the board and returns the first cell with valid wall parents
  // that it finds. If it doesn't find anything, returns -1
  //===========================================================================
  int Maze::getCellWithWallParent (bool acceptFinish) {
    int foundCell = -1;

    for (int i = 0; i < _totalSize && foundCell < 0; i++) {
      int tmpCell = getRandomWallParent (i, acceptFinish);
      if (tmpCell > -1) {
        foundCell = i;
      }
    }

    return foundCell;
  }


  //===========================================================================
  // getRandomWallParent
  // 
  // Receives a cell and chooses 1 random parent that is a wall
  // to return (only valid wall parents are returned, if there are no valid
  // parents, the function returns -1)
  //===========================================================================
  int Maze::getRandomWallParent (int cell, bool acceptFinish) {
    int* validParents = getValidParents (cell);

    // Counting how many parents are valid, if none of them is valid, returns
    // -1
    int countValidParents = 0;
    for (int i = 0; i < 4; i++) {
      bool allowedParent = _board [validParents [i]] == BOARD_WALL;

      if (acceptFinish) {
        allowedParent = allowedParent ||
          validParents [i] == _totalSize -1;
      }
      

      if (validParents [i] > -1 && allowedParent) {
        countValidParents++;
      }
    }
    if (countValidParents < 1) return -1;

    // Looping until it finds a random parent that is valid
    bool isValidParent = false;
    int selectedNumber = -1;
    do {
      int randomNumber = std::rand () % 4;
      bool allowedParent = _board [validParents [randomNumber]] == BOARD_WALL;

      if (acceptFinish) {
        allowedParent = allowedParent ||
          validParents [randomNumber] == _totalSize - 1;
      }

      isValidParent = (validParents [randomNumber] > -1 && allowedParent);
      if (isValidParent) selectedNumber = validParents [randomNumber];
    } while (!isValidParent);
    
    delete validParents;
    return selectedNumber;
  }


  //===========================================================================
  // getValidParents
  // 
  // Receives a cell and returns an array with 8 elements containing the cell's
  // valid parents. If the parent is not valid, its index will be -1
  //===========================================================================
  int* Maze::getValidParents (int cell) {
    int* parentCells = new int [4];

    int currentRow = getRow (cell);

    // top
    parentCells [0] = cell - _size;

    // right
    parentCells [1] = cell + 1;

    // bottom
    parentCells [2] = cell + _size;

    // left
    parentCells [3] = cell - 1;

    // Checking if the right/left cells are not jumping to another row
    bool overlappingRowRight = getRow (parentCells [1]) != currentRow;
    bool overlappingRowLeft  = getRow (parentCells [3]) != currentRow;

    // Getting only valid parents, the other ones are set to -1
    for (int i = 0; i < 4; i++) {
      if (parentCells [i] < 0 ||
          parentCells [i] >= _totalSize ||
          (overlappingRowRight && i == 1) ||
          (overlappingRowLeft  && i == 3)) {
        parentCells [i] = -1;
      }
    }

    return parentCells;
  }


  //===========================================================================
  // getRow
  // 
  // Receives a cell and returns the index of the row this cell belongs to
  //===========================================================================
  int Maze::getRow (int cell) {
    return cell / _size;
  }


  //===========================================================================
  // getCol
  // 
  // Receives a cell and returns the index of the column this cell belongs to
  //===========================================================================
  int Maze::getCol (int cell) {
    return cell % _size;
  }


  //===========================================================================
  // display
  // 
  // Receives an ostream, records the board in the stream and then returns it.
  // Can be accessed through the operator <<
  //===========================================================================
  std::ostream& Maze::display (std::ostream& out) {
    // Making an exit
    pushLink (_totalSize -1, (_totalSize - 1) + _size);

    // Upper wall
    for (int i = 0; i < (_size * 2) + 1; i++) {
      if (i == 1) {
        out << BOARD_EMPTY;
      } else {
        out << BOARD_WALL;
      }
    }
    out << std::endl;

    // Holds the bottom wall for the rows
    string bottomWall = "";

    for (int i = 0; i < _totalSize; i++) {
      // Printing left border if needed
      if (getCol (i) == 0) {
        out << BOARD_WALL;
        bottomWall += BOARD_WALL;
      }

      // Printing the cell
      out << _board [i];

      // If there is a link between the cells vertically, prints an empty
      // path, otherwise, a wall in the bottom wall
      if (hasLink (i, i + _size)) {
        bottomWall += BOARD_EMPTY;
      } else {
        bottomWall += BOARD_WALL;
      }

      // If there is a link between the cells horizontally, prints an empty
      // path, otherwise, a wall
      if (hasLink (i, i + 1)) {
        out << BOARD_EMPTY;
      } else {
        out << BOARD_WALL;
      }
      bottomWall += BOARD_WALL;

      // Finished printing the line, prints the right wall and the row below
      // with the maze walls
      if (getCol (i) == (_size - 1)) {
        //out << BOARD_WALL << std::endl;
        //bottomWall += BOARD_WALL;

        out << std::endl << bottomWall << std::endl;

        bottomWall = "";
      }
    }

    // Bottom wall
    //for (int i = 0; i < (_size * 2 + 1); out << 'o', i++);
    //out << std::endl;

    return out;
  }


  //===========================================================================
  // <<
  // 
  // Receives an ostream, records the board in the stream and then returns it.
  //===========================================================================
  std::ostream& operator << (std::ostream& out, Maze& board) {
    return board.display (out);
  }
}

File: Maze-Generator-Java/mazeGenerator/Maze.java
package mazeGenerator;

import java.util.*;

/**
 * Board
 *
 * Represents a Maze
 *
 * @author Henrique Salvadori Coelho - henriquesc@gmail.com - hcoelho.com
 */

public class Maze {

  // Charactes for the blocks in the maze
  public static final char BOARD_WALL   = '#';
  public static final char BOARD_EMPTY  = ' ';
  public static final char BOARD_START  = ' ';
  public static final char BOARD_FINISH = ' ';

  private final Random seed;
  private final int size;
  private final int totalSize;
  private final int startPoint;
  private final int finishPoint;

  private char[] board;
  private HashSet<Integer>[] links; // Child-parent between cells

  public Maze(int size) {
    if (size > 101 || size < 7 || size % 2 == 0) {
      throw new IllegalArgumentException ("Board size not correctly set. It must be an odd number between 7 and 101.");
    }

    this.size = size / 2; // The real size (width and height) are only half of the board, the rest are walls
    this.totalSize = this.size * this.size; // Width * Height
    this.links = new HashSet[totalSize + size];
    this.startPoint = 0;
    this.finishPoint = totalSize - 1;

    // Initializing the board. The loop goes to an additional row at the end to initialize the "links" hashset, in order
    // to create the finish point in the last row, but it is not recorded in the "board" array.
    this.board = new char[totalSize];
    for (int i = 0; i < totalSize + size; i++) {
      links[i] = new HashSet();
      if (i < totalSize) { board[i] = BOARD_WALL; }
    }

    // Placing the start and finish point
    board[startPoint]  = BOARD_START;
    board[finishPoint] = BOARD_FINISH;

    this.seed = new Random();

    // Building the maze
    buildMaze();
  }

  /**
   * This method is responsible for building the maze. The maze starts as a "solid block" of walls, and this algorithm
   * will create paths through the maze until there are no available paths left between the walls
   */
  private void buildMaze() {
    boolean foundFinish = false;

    // Looping while there still are breakable walls in the maze
    do {

      boolean hasAvailableCell;

      // Getting a cell that contains a wall. If the finish cell has not been reached yet, it can also count as a wall
      int currentCell = getCellWithWall(!foundFinish);

      // Looping to expand the path until it cannot be expanded anymore
      do {
        int selectedCell = getRandomWall(currentCell, !foundFinish);

        // If there is an available cell next to the current one, jumps to it
        hasAvailableCell = selectedCell > -1;
        if (hasAvailableCell) {

          // If it reached the finish, stops this path and starts a new one
          if (selectedCell == totalSize - 1) {
            foundFinish = true;
            hasAvailableCell = false;
          } else {
            board[selectedCell] = BOARD_EMPTY;
          }

          // Creating links between the cells
          links[currentCell] .add(selectedCell);
          links[selectedCell].add(currentCell);

          currentCell = selectedCell;
        }

      } while (hasAvailableCell);

    } while (getCellWithWall(!foundFinish) > -1);
  }

  /**
   * Loops through the board and returns the first cell with valid wall parents that it finds. If it doesn't find
   * anything, returns -1.
   *
   * @param acceptFinish If true, the finish cell is treated as a wall and can be returned as a valid cell
   * @return A random neighbor cell
   */
  private int getCellWithWall(boolean acceptFinish) {
    int foundCell = -1;

    for (int i = 0; i < totalSize; i++) {
      int tmpCell = getRandomWall(i, acceptFinish);
      if (tmpCell > -1) {
        foundCell = i;
        break;
      }
    }

    return foundCell;
  }

  /**
   * Receives a cell and chooses 1 random parent that is a wall to return (only valid wall parents are returned, if
   * there are no valid parents, the function returns -1)
   *
   * @param cell A cell which should contain neighbor walls
   * @param acceptFinish If true, the finish cell is treated as a wall and can be returned as a valid cell
   * @return A random neighbor that is a wall
   */
  private int getRandomWall(int cell, boolean acceptFinish) {
    List<Integer> validParents = new LinkedList<>();

    int currentRow = getRow(cell);

    // Getting all the possible neighbors. If a cell is invalid (out of bounds or overflowing row), the value is -1
    int cellTop    = cell - size;
    int cellRight  = getRow(cell + 1) == currentRow ? cell + 1 : -1;
    int cellBottom = cell + size;
    int cellLeft   = getRow(cell - 1) == currentRow ? cell - 1 : -1;

    // If the parents are valid, adds them to the list
    if (isValidWallOrFinish(cellTop   , acceptFinish)) { validParents.add(cellTop   ); }
    if (isValidWallOrFinish(cellRight , acceptFinish)) { validParents.add(cellRight ); }
    if (isValidWallOrFinish(cellBottom, acceptFinish)) { validParents.add(cellBottom); }
    if (isValidWallOrFinish(cellLeft  , acceptFinish)) { validParents.add(cellLeft  ); }

    // If none parent is valid, returns -1
    return validParents.isEmpty() ? -1 : validParents.get(Math.abs(seed.nextInt() + cell) % validParents.size());
  }

  /**
   * Helper for getRandomWall. This method receives a position of a cell and verifies if it is a wall or not
   *
   * @param cellPosition Position of the cell
   * @param acceptFinish Treat the finish cell as a wall or not
   * @return True if the cell is a valid wall, false otherwise
   */
  private boolean isValidWallOrFinish(int cellPosition, boolean acceptFinish) {
    if (cellPosition >= 0 && cellPosition < totalSize) {
      char cellValue = board[cellPosition];
      if (acceptFinish) {
        return (cellValue == BOARD_WALL || cellPosition == finishPoint) && cellPosition != startPoint;
      } else {
        return cellValue == BOARD_WALL && cellPosition != startPoint && cellPosition != finishPoint;
      }
    } else {
      return false;
    }
  }

  /**
   * Converts the position of a cell to the index of the row it belongs to
   *
   * @param cell The position of the cell to be converted
   * @return The row the cell belongs to
   */
  private int getRow(int cell) {
    return cell / size;
  }

  /**
   * Converts the position of a cell to the index of the column it belongs to
   *
   * @param cell The position of the cell to be converted
   * @return The column the cell belongs to
   */
  private int getCol(int cell) {
    return cell % size;
  }

  /**
   * Verifies if there is a link between two cells (if there is no link, there's a wall between them)
   *
   * @param cell1 First cell
   * @param cell2 Second cell
   * @return True if they are linked, false otherwise
   */
  private boolean hasLink(int cell1, int cell2) {
    if (links[cell1].contains(cell2)) { return true; }
    return false;
  }

  /**
   * Returns the size (width or height) of the maze
   *
   * @return An integer with the size of the maze
   */
  public int size() {
    return size;
  }

  /**
   * Allows the maze to be printed
   *
   * @return A string that represents the state of the object
   */
  @Override
  public String toString() {
    StringBuilder out = new StringBuilder();

    // Making an exit
    int a = totalSize - 1;
    int b = (totalSize - 1) + size;
    links[a].add(b);
    links[b].add(a);

    // Upper wall
    for (int i = 0; i < (size * 2) + 1; i++) {
      if (i == 1) { out.append(BOARD_EMPTY); }
      else        { out.append(BOARD_WALL); }
    }
    out.append(System.lineSeparator());

    // Holds the bottom wall for the rows
    StringBuilder bottomWall = new StringBuilder();

    for (int i = 0; i < totalSize; i++) {
      // Printing left border if needed
      if (getCol (i) == 0) {
        out.append(BOARD_WALL);
        bottomWall.append(BOARD_WALL);
      }

      // Printing the cell
      out.append(board[i]);

      // If there is a link between the cells vertically, prints an empty
      // path, otherwise, a wall in the bottom wall
      if (hasLink (i, i + size)) { bottomWall.append(BOARD_EMPTY); }
      else                       { bottomWall.append(BOARD_WALL); }

      // If there is a link between the cells horizontally, prints an empty
      // path, otherwise, a wall
      if (hasLink (i, i + 1)) { out.append(BOARD_EMPTY); }
      else                    { out.append(BOARD_WALL); }
      bottomWall.append(BOARD_WALL);

      // Finished printing the line, prints the right wall and the row below
      // with the maze walls
      if (getCol (i) == (size - 1)) {
        out.append(System.lineSeparator()).append(bottomWall.toString()).append(System.lineSeparator());
        bottomWall = new StringBuilder();
      }
    }

    return out.toString();
  }
}
